sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -152,3 +153,56 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
       if (j.joinType == newJoinType) f else Filter(condition, j.copy(joinType = newJoinType))
   }
 }
+
+/**
+ * Correctly handle PythonUDF which need access both side of join side by changing the new join
+ * type to Cross.
+ */
+object HandlePythonUDFInJoinCondition extends Rule[LogicalPlan] with PredicateHelper {
+  def hasPythonUDF(expression: Expression): Boolean = {
+    expression.collectFirst { case udf: PythonUDF => udf }.isDefined
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case j @ Join(_, _, joinType, condition)
+      if condition.map(splitConjunctivePredicates).getOrElse(Nil).exists(hasPythonUDF) =>
```

Comment:
why not just `if hasPythonUDF(condition)`?

Reply:
ah yes, thanks, done in 98cd3cc.

Reply:
@xuanyuanking So here we are finding out if the join condition has a python UDF. I am trying to understand where we are making the determination that this python UDF is referring to attributes of both legs of the join ? Can you please let me know. Thank you.

Reply:
@xuanyuanking Oh..  it is because the UDFS referring to single leg would have been pushed down and we will only have UDFs referring to both legs in the join condition when we come here ?

Reply:
@cloud-fan Just saw your comment that validated my understanding on this rule relying on pushing down predicates through join. However, the pushdownPredicateThroughJoin is not in the nonExcludableRules list. So can we rely on this rule being fired always here ?

Reply:
```
@xuanyuanking Oh.. is it because the UDFS referring to single leg would have been pushed down and we will only have UDFs referring to both legs in the join condition when we come here ?
```
Yea that's right.